bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreClientImpl.java

```diff
@@ -0,0 +1,581 @@
+
+      HttpResponse response =
+          client
+              .tables()
+              .delete(
+                  tableReference.getProjectId(),
+                  tableReference.getDatasetId(),
+                  tableReference.getTableId())
+              .executeUnparsed();
+
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchTableException(response.getStatusMessage());
+      }
+      convertExceptionIfUnsuccessful(response);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @Override
+  @SuppressWarnings("FormatStringAnnotation")
+  public List<Tables> listTables(
+      DatasetReference datasetReference, boolean filterUnsupportedTables) {
+    try {
+      String nextPageToken = null;
+      Stream<Tables> tablesStream = Stream.empty();
+      do {
+        HttpResponse pageResponse =
+            client
+                .tables()
+                .list(datasetReference.getProjectId(), datasetReference.getDatasetId())
+                .setPageToken(nextPageToken)
+                .executeUnparsed();
+        if (pageResponse.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+          throw new NoSuchNamespaceException(pageResponse.getStatusMessage());
+        }
+        TableList result = convertExceptionIfUnsuccessful(pageResponse).parseAs(TableList.class);
+        nextPageToken = result.getNextPageToken();
+        List<Tables> tablesPage = result.getTables();
+        Stream<Tables> tablesPageStream =
+            tablesPage == null ? Stream.empty() : result.getTables().stream();
+        tablesStream = Stream.concat(tablesStream, tablesPageStream);
+      } while (nextPageToken != null && !nextPageToken.isEmpty());
+
+      // The server should return more metadata here (e.g. BigQuery Non Iceberg tables) to
+      // distinguish Iceberg
+      // tables for us to filter out those results since invoking `getTable` on them would
+      // correctly raise a `NoSuchIcebergTableException` for being inoperable by this plugin.
+      if (filterUnsupportedTables) {
+        tablesStream =
+            tablesStream
+                .parallel()
+                .filter(
+                    table -> {
+                      try {
+                        getTable(table.getTableReference());
+                      } catch (NoSuchTableException e) {
+                        return false;
+                      }
+                      return true;
+                    });
+      }
+
+      return tablesStream.collect(Collectors.toList());
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  private Dataset updateDataset(Dataset dataset) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .update(
+                  Preconditions.checkNotNull(dataset.getDatasetReference()).getProjectId(),
```

Comment:
IMO this should be rather `Preconditions.checkArgument(.. != null, errorMsg)` as it is clearer for a reader of the error msg to know why something isn't a valid argument.